f3_data_builder.py

```diff
@@ -1,249 +1,632 @@
+    the table column_name / value_to_insert.
+    :type insert_values: list[dict[str, Any]]
+    :rtype: Insert[Any]
+    :return: SQLAlchemy INSERT statement
+    """
+    sql = insert(table).values(insert_values)
+    on_dup = sql.on_duplicate_key_update({v.name: v for v in sql.inserted if v.name in update_cols})
+    return on_dup
+
+
+def region_subquery(metadata: MetaData) -> Subquery:
+    """
+    Abstracting some SQL duplication between the
+    paxminer and weaselbot region queries
+    """
+    cb = metadata.tables["weaselbot.combined_beatdowns"]
+    a = metadata.tables["weaselbot.combined_aos"]
+
+    sql = select(
+        a.c.region_id,
+        func.max(cb.c.timestamp).label("max_timestamp"),
+        func.max(cb.c.ts_edited).label("max_ts_edited"),
+        func.count().label("beatdown_count"),
+    )
+    sql = sql.select_from(cb.join(a, cb.c.ao_id == a.c.ao_id))
+    sql = sql.group_by(a.c.region_id).subquery("b")
+    return sql
+
+
+def paxminer_region_query(metadata: MetaData, cr: Table) -> Selectable:
+    """
+    Construct the region SQL using paxminer
+    """
+    r = metadata.tables["paxminer.regions"]
+    sub = region_subquery(metadata)
+
+    sql = select(
+        r.c.schema_name,
+        r.c.region.label("region_name"),
+        sub.c.max_timestamp,
+        sub.c.max_ts_edited,
+        sub.c.beatdown_count,
+        cr.c.region_id,
+    )
+    sql = sql.select_from(
+        r.outerjoin(cr, r.c.schema_name == cr.c.schema_name).outerjoin(sub, cr.c.region_id == sub.c.region_id)
+    )
+
+    return sql
+
+
+def weaselbot_region_query(metadata: MetaData, cr: Table) -> Selectable:
+    """
+    Construct the region SQL using weaselbot
+    """
+    sub = region_subquery(metadata)
+
+    sql = select(cr, sub.c.beatdown_count)
+    sql = sql.select_from(cr.outerjoin(sub, cr.c.region_id == sub.c.region_id))
+
+    return sql
+
+
+def region_queries(engine: Engine, metadata: MetaData) -> pd.DataFrame:
+    """
+    Using PAXMiner and Weaselbot region tables, make updates to the
+    Weaselbot combined_regions table if any exist from PAXMiner.
+
+    :param engine: SQLAlchemy connection engine to MySQL
+    :type engine: sqlalchemy.engine.Engine object
+    :param metadata: collection of reflected table metadata
+    :type metadata: SQLAlchemy MetaData
+    :rtype: pandas.DataFrame
+    :return: A dataframe containing current region information
+    """
+    cr = metadata.tables["weaselbot.combined_regions"]
+
+    paxminer_region_sql = paxminer_region_query(metadata, cr)
+
+    df_regions = pd.read_sql(paxminer_region_sql, engine)
+    insert_values = df_regions.to_dict("records")
+    update_cols = ("region_name", "max_timestamp", "max_ts_edited")
+    region_insert_sql = insert_statement(cr, insert_values, update_cols)
+
+    with engine.begin() as cnxn:
+        cnxn.execute(region_insert_sql)
+
+    dtypes = dict(
+        region_id=pd.StringDtype(),
```

Comment:
Regarding the `id` column data types, I noticed that among the various tables, they alternate between `int` and `string`. I figured, for the sake of consistency and to avoid any weird trailing `.0`'s, I'd cast _all_ `id` columns to `string`. In `pandas`, if the merge columns aren't all the same data type, then joins don't work. So I figured strings across the board. Happy to discuss.